### DISCUSSION TITLE
What is the best practice to "manage" structs?

### DISCUSSION THREAD

**Discussion by JSFTI** [diesel-rs/diesel#3770] [Q&A] [answered]
I have an entity that has multiple nested relations. So, I created multiple structs, each has different relation attributes. For example:
```
Post  
PostWithComments  
PostWithCommentsUser  
UserWithPosts  
```
To get these data from the database, I created multiple functions such as:
```
// posts.rs
pub fn find(post_id: i32) -> Post
pub fn get_with_comments() -> Vec<PostWithComments>
pub fn get_with_comments_user() -> Vec<PostWithCommentsUser>

// users.rs
pub fn find_with_posts(user_id: i32) -> UserWithPosts
```

This methods "works". I got the data structure that I want and is able to be sent as JSON (with serde). However, this is getting too painful to manage. The database has around 15 tables. Some of them have multiple deeply nested relations. All of the relations of the entity have to be sent.

I am quite new in Rust, never had a professional project done in Rust. I came from Laravel, and the models are easier to manage because PHP is dynamic typed and the Eloquent ORM is very powerful. With my current method, creating an API in Rust feels more difficult and slow (in writing code) than in Laravel. So, I was wondering if there is an easier method in managing data models in Rust + Diesel.

**Comment by weiznich** [comment, marked as answer]
It's hard to give a general answer on such questions, especially without knowing how you structured your structs in detail. It really depends on your exact use case and which level of composability and type safety you want. 
Nerveless let's try to provide at least some answers:

The first point  to notice is that if you have different json structures that you want to return via serde you need to either have different rust structures for that (in combination with `#[derive(Serialize)]`) or use `serde_json::json!` to construct the relevant json ad hoc. Which variant is better really depends on your needs there. That's mostly due to the statically typed nature of rust. If you have types you need to provide them for everything. I would probably use the former variant for safety and only use the later variant for one-of things.

On diesel side the relevant parts to read are [the associations](https://docs.diesel.rs/2.1.x/diesel/associations/index.html) documentation and [the relations guide](http://diesel.rs/guides/relations.html). It seems like all your use-cases are essentially a 1:n relation thing, which is mostly covered by the associations documentation. (Note that this also allows arbitrarily deep nesting) The main point there is that you somewhat get a `Vec<(User, Vec<Post>)>` data structure out of that. Now combine that with the things written about serde above you have several options:
* Iterate over it and construct the json via `serde_json::json!()`:
```rust
results.into_iter().map(|(user, posts)| {
    serde_json::json!({
        "name": user.name,
        "…": user.other_field,
        "posts": posts,
}).collect()
```
* Create intermediate struct to put `#[derive(Serialize)]` onto:
```rust
#[derive(Serialize)]
struct UserWithPosts {
    #[serde(flatten)] // this part allows us to skip duplicating all the fields and do mapping there
    user: User,
    posts: Vec<Post>,
}

results.into_iter().map(|(user, posts)| UserWithPosts {user, posts}).collect();
```

Both variants will result in a json like that:
```json
{
    "name": "username",
    "posts": [{"title": "post title"}]
}
```

I hope that helps at least a bit.

**Reply by JSFTI** [reply]
I apologize for not giving my defined structs in detail. I was going to copy paste my whole models in here, but I was worried that it might just be more confusing since the code is quite long and there are a lot of models. So I just write a general example of my problem.

Your answer really helped. I never realized that serde_json has a macro to create an "object" without defining a struct first. Though the trade-off is the compiler is not able to tell the attributes of the "object" anymore, increasing chances for runtime errors. I think this is similar with Java's Object. Having flexibility but sacrificing type safety. 

I guess my future question is also answered. I am confused in creating APIs that allows freedom in choosing fields to return. Kind of like GraphQL.

For example:
```
// ?select=name,age,dob

{
  name: 'John Smith',
  age: 21,
  dob: '2001-04-20'
}
```
```
// ?select=name,age

{
  name: 'John Smith',
  age: 21
}
```

With Rust's strict type system, I guess this is impossible to do without something like serde_json::value::Value. Can't imagine creating _n_ number of structs with _n_ as the combination of possible selected fields.

I think that your example with UserWithPosts kind of "validated" my current solution. I am already using `#[serde(flatten)]`. I was just hoping that maybe there exists an easier solution like in dynamic languages (like PHP) but is still type safe.

I have already used associations to help in querying relational data. The Diesel guide really helped.

I guess my question is answered. Thank you for the help.

**Reply by weiznich** [reply]
> With Rust's strict type system, I guess this is impossible to do without something like serde_json::value::Value. Can't imagine creating n number of structs with n as the combination of possible selected fields.
>
> I think that your example with UserWithPosts kind of "validated" my current solution. I am already using #[serde(flatten)]. I was just hoping that maybe there exists an easier solution like in dynamic languages (like PHP) but is still type safe.

There is another solution to this kind of problem: You can make all fields in your struct used for serialization `Option<Whatever>` and configure serde in such a way to skip `None` fields entirely. On database side you would need to generate queries that "just" return `NULL` values for such fields if that's requested by the input query to avoid loading unnecessary data.

**Reply by JSFTI** [reply]
That's actually a pretty nice solution. I didn't know serde is capable of that. I should start reading documentations more thoroughly. Though that quotation marks on the "just" word is pretty worrying. However, this might be a good exercise for me.

Thank you for the solution.

**Reply by weiznich** [reply]
"just" means in that context that there is no diesel built-in primitive for that nor is there a guide how to do that. 
In the end you can likely use a pattern like that:
```rust
let name = if select_name { Box::new(users::name.nullable()) as Box<dyn BoxableExpression<users::table, _, SqlType = Nullable<Text>>>} else { Box::new(None::<String>.into_sql::<Nullable<Text>>()) };
// similar statements for the other fields

// at your query
query.select((name, age, …))

```

**Reply by JSFTI** [reply]
I see... I am going to bookmark that. My current project does not need "selectable" fields yet. I was just asking out of curiosity.

Thank you so much for your help. I have been thinking about ways to structure Rust code base for building APIs and data models for a while. Your advices really helped in clearing some concepts for me.
